Deprecate legacy input parameter construction (#19448)
idaholab/moose: modules/doc/content/newsletter/2021_11.md

```

And the parameters were defined in the source by specializing
`InputParameters validParams<T>()` as follows:

```cpp
template <>
InputParameters
validParams<SomeMooseObject>()
{
  InputParameters params = validParams<SomeParentMooseObject>();
  params.addParam<bool>("foo", false, "Some description");
  return params;
}
```

The new method of input parameter construction involves defining
a static method on the object itself: `static InputParameters validParams()`.
The transition of objects to the new input parameter construction method
is straight-forward and will be demonstrated using the example object above.
First, replace all calls to `validParams<T>()` with `T::validParams()`.

In the header of each object, remove the forward declarations as
described above and add a public `static InputParameters validParams()`
declaration:

```cpp
#include "SomeParentMooseObject.h"

class SomeMooseObject : public SomeParentMooseObject
{
public:
  SomeMooseObject(const InputParameters & params);

  static InputParameters validParams();

  // continued...
};
```

In the source of each object, replace the template specialization of
`InputParameters validParams<T>()` with the

> `T::InputParameters validParams()` method:

Author: ```suggestion
`InputParameters T::validParams()` method:
```

If any issues with the new packages are experienced, please open a new
[Discussions post](https://github.com/idaholab/moose/discussions)!

## Legacy Input Parameter Deprecation

The legacy input parameter construction method has been supported for
nearly two years. It is now slated for deprecation on Mar 1, 2022.
After said date, any objects whose parameter construction is not
updated will not compile.

> Previously, input parameters were defined with a template specialization
> of `InputParameters validParams<T>()`.
> 
> This required a forward declaration in the header for an object as follows:

Reviewer: ```suggestion
Previously, input parameters were defined with a template specialization
of `InputParameters validParams<T>()`, which required a forward declaration in the object header as follows:
```

```cpp
class SomeMooseObject;
template <>
InputParameters validParams<SomeMooseObject>
```

And the parameters were defined in the source by specializing
`InputParameters validParams<T>()` as follows:

```cpp
template <>
InputParameters
validParams<SomeMooseObject>()
{
  InputParameters params = validParams<SomeParentMooseObject>();
  params.addParam<bool>("foo", false, "Some description");
  return params;
}
```

The new method of input parameter construction involves defining
a static method on the object itself: `static InputParameters validParams()`.
The transition of objects to the new input parameter construction method

> is straight-forward (taking the example object above). First, replace all

Reviewer: ```suggestion
is straight-forward and will be demonstrated using the example object above. First, replace all
```

If any issues with the new packages are experienced, please open a new
[Discussions post](https://github.com/idaholab/moose/discussions)!

## Legacy Input Parameter Deprecation

The legacy input parameter construction method has been supported for
nearly two years. It is now slated for deprecation on Mar 1, 2022.
After said date, any objects whose parameter construction is not
updated will not compile.

Previously, input parameters were defined with a template specialization
of `InputParameters validParams<T>()`, which required a forward declaration
in the object header as follows:

> 
> This required a forward declaration in the header for an object as follows:

Reviewer: ```suggestion
```

This got left in when you accepted my review.
